File: c++/Lexicographical Numbers.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<int> lexicalOrder(int n) {
        std::vector<int> result;
        int current = 1;
        for (int i = 0; i < n; ++i) {
            result.push_back(current);
            if (current * 10 <= n) {
                current *= 10;
            } else {
                if (current >= n) {
                    current /= 10;
                }
                current++;
                while (current % 10 == 0) {
                    current /= 10;
                }
            }
        }
        return result;
    }
};

// Test cases
int main() {
    Solution solver;

    int n1 = 13;
    std::vector<int> result1 = solver.Solution(n1);
    std::cout << "Lexicographical order for n = " << n1 << ": ";
    for (int num : result1) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    // case 2
    int n2 = 2;
    std::vector<int> result2 = solver.Solution(n2);
    std::cout << "Lexicographical order for n = " << n2 << ": ";
    for (int num : result2) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    //  case 3
    int n3 = 100;
    std::vector<int> result3 = solver.Solution(n3);
    std::cout << "Lexicographical order for n = " << n3 << ": ";
    for (int num : result3) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: python/Lexicographical Numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> list[int]:
        result = []
        current = 1
        
        for _ in range(n):
            result.append(current)
            if current * 10 <= n:
                current *= 10
            else:
                if current >= n:
                    current //= 10
                current += 1
                while current % 10 == 0:
                    current //= 10
                    
        return result

# Test cases
if __name__ == "__main__":
    solution = Solution()

    # case 1
    print(f"result of test case 1 is : {solution.lexicalOrder(13)}") # Expected: [1,10,11,12,13,2,3,4,5,6,7,8,9]

    # case 2
    print(f"result of test case 2 is : {solution.lexicalOrder(2)}") # Expected: [1,2]
